### PR TITLE
K8s - redeployment

### PR DESCRIPTION
**Pull request by mproch**

### CODE REVIEW

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 1]
```diff
@@ -168,7 +112,79 @@ class K8sDeploymentManager(modelData: ModelData, config: K8sDeploymentManagerCon
 
   override def findJobStatus(name: ProcessName): Future[Option[ProcessState]] = {
     val mapper = new K8sDeploymentStatusMapper(processStateDefinitionManager)
-    k8s.listSelected[ListResource[Deployment]](labelSelectorForName(name)).map(_.items).map(mapper.findStatusForDeployments)
+    k8s.listSelected[ListResource[Deployment]](requirementForName(name)).map(_.items).map(mapper.findStatusForDeployments)
+  }
+
+  //TODO: use https://kubernetes.io/docs/reference/using-api/server-side-apply/ in the future
+  private def createOrUpdate[O<:ObjectResource](data: O)
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:119]: Can u move it to `K8sUtils`?

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 2]
```diff
@@ -168,7 +112,79 @@ class K8sDeploymentManager(modelData: ModelData, config: K8sDeploymentManagerCon
...
+                                               (implicit fmt: Format[O], rd: ResourceDefinition[O], lc: LoggingContext): Future[O] = {
+    k8s.getOption[O](data.name).flatMap {
+      case Some(_) => k8s.update(data)
+      case None => k8s.create(data)
+    }
+  }
+
+  protected def configMapForData(processVersion: ProcessVersion, deploymentData: ProcessDeploymentData): ConfigMap = {
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:127]: I wonder if we could extract some method like `K8sUtils.appendHash(configMap, inputForHash)` to not obfuscate logic of our deployment with k8s stuff. WDYT?

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 3]
```diff
@@ -189,31 +205,48 @@ object K8sDeploymentManager {
...
+  private[manager] def objectNameForScenario(processVersion: ProcessVersion, append: String = ""): String = {
+    sanitizeObjectName(s"scenario-${processVersion.processId.value}-${processVersion.processName.value}", append)
   }
 
   private[manager] def parseVersionAnnotation(deployment: Deployment): Option[ProcessVersion] = {
     deployment.metadata.annotations.get(scenarioVersionAnnotation).flatMap(CirceUtil.decodeJson[ProcessVersion](_).toOption)
   }
 
+  //https://github.com/kubernetes/kubectl/blob/master/pkg/util/hash/hash.go#L105 - we don't care about bad words...
+  private[manager] def shortHash(data: String): String = DigestUtils.sha256Hex(data).take(10)
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:250]: Can u move it to `K8sUtils`?

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 4]
```diff
@@ -189,31 +205,48 @@ object K8sDeploymentManager {
...
 
   /*
     Id of both is created with scenario id + sanitized name. This is to:
     - guarantee uniqueness - id is sufficient for that, sanitized name - not necessarily, as replacement/shortening may lead to duplicates
       (other way to mitigate this would be to generate some hash, but it's a bit more complex...)
     - ensure some level of readability - only id would be hard to match name to scenario
    */
-  private[manager] def objectNameForScenario(processVersion: ProcessVersion): String = {
-    sanitizeObjectName(s"scenario-${processVersion.processId.value}-${processVersion.processName}")
+  private[manager] def objectNameForScenario(processVersion: ProcessVersion, append: String = ""): String = {
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:241]: Would be nice to have some unit test checking if for long process name with the same prefix but other hash is generated different object name. I know that you have done it correctly but it is easy to be broken and unit test looks like simple thing to add

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala (hunk 1 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala]: :+1:

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala (hunk 3 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala]: :+1:

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala (hunk 2 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:115]: I refactored a bit but didn't create such a method, as id generation is a bit specific anyway. Look if it's ok for you now

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala (hunk 4 above)
Comment (mproch) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala]: :+1: added some tests

File: engine/lite/k8sDeploymentManager/src/test/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManagerUnitTest.scala [hunk 5]
```diff
@@ -0,0 +1,51 @@
...
+
+    val names = Table(
+      ("scenario name", "hashInput", "object id"),
+      ("standard", None, "scenario-256-standard"),
+      ("ała", None, "scenario-256-a-a"),
+      //here we don't care about hash, as id should be unique
+      ("a-a", None, "scenario-256-a-a"),
+      //but for different content, id should be different
+      ("st", Some("version1"), "scenario-256-st-fc6af3ec64"),
+      ("st", Some("version2"), "scenario-256-st-366a963b60"),
```
Comment (arkadius) [inline engine/lite/k8sDeploymentManager/src/test/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManagerUnitTest.scala:41]: Can u add two test cases wit scenario name overflowing limit of 63 characters but different hashes? it was the case that was most important for me

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala (hunk 2 above)
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala:115]: It's ok

File: engine/lite/k8sDeploymentManager/src/test/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManagerUnitTest.scala (hunk 5 above)
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/test/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManagerUnitTest.scala:62]: thanks

File: engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala [hunk 6]
```diff
@@ -72,10 +69,10 @@ class K8sDeploymentManager(modelData: ModelData, config: K8sDeploymentManagerCon
                       processDeploymentData: ProcessDeploymentData,
                       savepointPath: Option[String]): Future[Option[ExternalDeploymentId]] = {
     for {
-      configMap <- createOrUpdate(configMapForData(processVersion, processDeploymentData))
+      configMap <- createOrUpdate(k8s, configMapForData(processVersion, processDeploymentData))
```
[the remark this answers is not in the document]
Comment (arkadius) [reply engine/lite/k8sDeploymentManager/src/main/scala/pl/touk/nussknacker/k8s/manager/K8sDeploymentManager.scala]: thanks

**Pull request merged**